frappe/core/page/background_jobs/background_jobs.py

```diff
@@ -1,58 +1,84 @@
 # Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
 # MIT License. See license.txt
 
-from __future__ import unicode_literals
-import frappe
-
+from typing import TYPE_CHECKING, Dict, List
 from rq import Queue, Worker
+
+import frappe
+from frappe import _
+from frappe.utils import cint, convert_utc_to_user_timezone, format_datetime
 from frappe.utils.background_jobs import get_redis_conn
-from frappe.utils import format_datetime, cint, convert_utc_to_user_timezone
 from frappe.utils.scheduler import is_scheduler_inactive
-from frappe import _
 
-colors = {
+if TYPE_CHECKING:
+	from rq.job import Job
+
+COLORS = {
 	'queued': 'orange',
 	'failed': 'red',
 	'started': 'blue',
 	'finished': 'green'
 }
 
+
 @frappe.whitelist()
-def get_info(show_failed=False):
+def get_info(show_failed=False) -> List[Dict]:
 	conn = get_redis_conn()
 	queues = Queue.all(conn)
 	workers = Worker.all(conn)
 	jobs = []
 
-	def add_job(j, name):
-		if j.kwargs.get('site')==frappe.local.site:
-			jobs.append({
-				'job_name': j.kwargs.get('kwargs', {}).get('playbook_method') \
-					or j.kwargs.get('kwargs', {}).get('job_type') \
-					or str(j.kwargs.get('job_name')),
-				'status': j.get_status(), 'queue': name,
-				'creation': format_datetime(convert_utc_to_user_timezone(j.created_at)),
-				'color': colors[j.get_status()]
-			})
-			if j.exc_info:
-				jobs[-1]['exc_info'] = j.exc_info
-
-	for w in workers:
-		j = w.get_current_job()
-		if j:
-			add_job(j, w.name)
-
-	for q in queues:
-		if q.name != 'failed':
-			for j in q.get_jobs(): add_job(j, q.name)
+	def add_job(job: 'Job', name: str) -> None:
+		if job.kwargs.get('site') == frappe.local.site:
+			job_info = {
+				'job_name': job.kwargs.get('kwargs', {}).get('playbook_method')
+					or job.kwargs.get('kwargs', {}).get('job_type')
+					or str(job.kwargs.get('job_name')),
+				'status': job.get_status(),
+				'queue': name,
+				'creation': format_datetime(convert_utc_to_user_timezone(job.created_at)),
+				'color': COLORS[job.get_status()]
+			}
+
+			if job.exc_info:
+				job_info['exc_info'] = job.exc_info
 
+			jobs.append(job_info)
+
+	# show worker jobs
+	for worker in workers:
+		job = worker.get_current_job()
+		if job:
+			add_job(job, worker.name)
+
+	# show active queued jobs
+	for queue in queues:
+		if queue.name != 'failed':
+			for job in queue.jobs:
+				add_job(job, queue.name)
+
+	# show failed jobs, if requested
 	if cint(show_failed):
-		for q in queues:
-			if q.name == 'failed':
-				for j in q.get_jobs()[:10]: add_job(j, q.name)
+		for queue in queues:
+			fail_registry = queue.failed_job_registry
+			for job_id in fail_registry.get_job_ids():
+				job = queue.fetch_job(job_id)
+				add_job(job, queue.name)
 
 	return jobs
 
+
+@frappe.whitelist()
+def remove_failed_jobs():
+	conn = get_redis_conn()
+	queues = Queue.all(conn)
+	for queue in queues:
+		fail_registry = queue.failed_job_registry
+		for job_id in fail_registry.get_job_ids():
```

Comment:
Can we add a test for this?

Reply:
Do I add the file in the page's folder? Will the tests be picked up from there?

Reply:
@Alchez You can add a new test file for background jobs under https://github.com/frappe/frappe/blob/develop/frappe/tests/

Reply:
@surajshetty3416, I've added a test for removing failed jobs

![image](https://user-images.githubusercontent.com/13396535/107929444-aa258180-6f9f-11eb-812f-6b848274a779.png)

Reply:
@surajshetty3416, Travis seems to be failing at enqueueing the failing function. I'm guessing that's a test environment issue. What do you suggest?

Reply:
@Alchez I was able to replicate the issue on my local, the test used to fail for the first time after bench start... seems like the enqueued job was not getting executed before the test ran... I've added `time.sleep` to wait for the job to get executed...